server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java

```diff
@@ -349,6 +351,97 @@ public Query rangeQuery(
                 context
             );
         }
+
+        public static class DateRangeDocValuesLoader extends BlockDocValuesReader.DocValuesBlockLoader {
+            private final String fieldName;
+
+            public DateRangeDocValuesLoader(String fieldName) {
+                this.fieldName = fieldName;
+            }
+
+            @Override
+            public Builder builder(BlockFactory factory, int expectedCount) {
+                return factory.longsFromDocValues(expectedCount);
+            }
+
+            @Override
+            public AllReader reader(LeafReaderContext context) throws IOException {
+                var docValues = context.reader().getBinaryDocValues(fieldName);
+                return new DateRangeDocValuesReader(docValues);
+                /*
+                if (docValues != null) {
+                    NumericDocValues singleton = DocValues.unwrapSingleton(docValues);
+                    if (singleton != null) {
+                        return new BlockDocValuesReader.SingletonLongs(singleton);
+                    }
+                    return new BlockDocValuesReader.Longs(docValues);
+                }
+                NumericDocValues singleton = context.reader().getNumericDocValues(fieldName);
+                if (singleton != null) {
+                    return new BlockDocValuesReader.SingletonLongs(singleton);
+                }
+                return new ConstantNullsReader(); */
+            }
+        }
+
+        @Override
+        public BlockLoader blockLoader(BlockLoaderContext blContext) {
+            if (hasDocValues()) {
+                return new DateRangeDocValuesLoader(name());
+            }
+            throw new IllegalStateException("Cannot load blocks without doc values");
+        }
+    }
+
+    public static class DateRangeDocValuesReader extends BlockDocValuesReader {
+        private final BytesRef spare = new BytesRef();
+
+        private final BinaryDocValues numericDocValues;
+
+        public DateRangeDocValuesReader(BinaryDocValues numericDocValues) {
+            this.numericDocValues = numericDocValues;
+        }
+
+        @Override
+        protected int docId() {
+            return 0;
```

Comment:
This'll need implementing. Without it we'll incorrectly try and reuse this when loading after a topn.